### ISSUE TITLE
AVG aggregate function not properly averaging

### ISSUE THREAD

**Issue by pruhstal** [sequelize/sequelize#9853]
## What are you doing?
Attempting to use the `avg` aggregate function within my query on a related `strain_reviews` table. 

```js
      const record = await Strain.find({
        where: {
          slug: req.params.slug
        },
        include: [
          {
            model: Category,
            where: {
              slug: req.params.type
            }
          },
          {
            model: StrainReview,
            where: {
              status: STATUS_TYPE_ACTIVE
            },
            include: [
              {
                model: User,
                attributes: ['id', 'photo', 'username']
              }
            ]
          },
          {
            model: StrainAttribute,
            include: [
              {
                model: Category,
                include: [{ model: Category, as: 'parent_category' }]
              }
            ]
          }
        ],
        attributes: {
          include: [
            [sequelize.fn('AVG', sequelize.col('strain_reviews.rating')), 'ratingAverage']
          ]
        },
        group: [
          'strain.id',
          'category.id',
          'strain_attributes.id',
          'strain_attributes->category.id',
          'strain_attributes->category->parent_category.id',
          'strain_reviews->user.id',
          'strain_reviews.id'
        ]
});
```

## What do you expect to happen?
I expected `ratingAverage` to be `3.5` (5 + 5 + 3 + 1) / 4

## What is actually happening?
The rating is returning the incorrect average. Looks like it just picks a random item in the array of `strain_reviews`.

_Output, either JSON or SQL_
(This is the object that gets returned with the non useful information removed):

```
{ id: 2,
  ...
  ratingAverage: '5.0000000000000000',
  strain_reviews:
   [ { id: 5,
       body: 'Gonna add my review now!',
       rating: 5,
       status: 1,
       strain_id: 2,
       user_id: 3,
       created_at: '2018-07-05T20:42:21.154Z',
       updated_at: '2018-07-05T20:42:21.154Z' },
     { id: 6,
       body: 'Gonna have to say yes!',
       rating: 5,
       status: 1,
       strain_id: 2,
       user_id: 5,
       created_at: '2018-07-05T20:42:51.447Z',
       updated_at: '2018-07-05T20:42:51.447Z' },
     { id: 4,
       body: 'Averaging this thang out!',
       rating: 1,
       status: 1,
       strain_id: 2,
       user_id: 2,
       created_at: '2018-06-27T21:39:22.690Z',
       updated_at: '2018-06-27T21:39:22.690Z' },
     { id: 9,
       body: 'Testing this out',
       rating: 3,
       status: 1,
       strain_id: 2,
       user_id: 1,
       created_at: '2018-07-22T19:27:23.198Z',
       updated_at: '2018-08-29T04:01:38.046Z' } ]
 }
```

__Dialect:__ postgres
__Database version:__ 10.4
__Sequelize version:__ 5.0.0-beta.11
__Tested with latest release:__ Yes

**Note :** _Your issue may be ignored OR closed by maintainers if it's not tested against latest version OR does not follow issue template._

**Comment by german970814** [comment]
Why this is closed? still happening on v5.8

**Comment by papb** [comment]
@german970814 Are you also using postgres?

**Comment by nikluwfy** [comment]
I do not claim big experience in this but as far as there is no other answers I will suggest my opinion:

You use group by option here, which basically looks for similarities in records and group them by specified ones. As I can see you group all by their ID's, which basically returns you all found records with their unique ID's but all of them are grouped and each group consists of solo record as long as IDs are unique. When you call aggregation functions they are being executed for the group scope, not for table scope. So basically try to group some records that have some common/similar data in their records and try to apply aggregation to them and see if it works. Looking forward hearing from you

**Comment by FilipeVeber** [comment]
I'm also facing this problem.
- If I run the SQL script directly against my database, the result is OK.
- If I run via my code, the result is wrong.

The difference between these both ways is that when the script is run via my code, it selects the "id" and the "serviceProviderId" (used in the group by clause) columns too while when I run it on my own, I remove them from the script.


This is my model association (ServiceProvider has many Rating):

```javascript
// ServiceProvider stuff...
include: [
      // other includes... 
	{
	  model: Rating,
	  as: `rating`,
	  separate: true,
	  attributes: [
		[
		  sequelize.fn(`ROUND`, sequelize.fn(`AVG`, sequelize.col(`rating`)), 2), `ratingAvg`,
		],
	  ],
	  where: { isFromUser: true },
	  group: [`serviceProviderId`],
	}
]
```

Below is the SQL output:

```SQL
[winston] Attempt to write logs with no transports {"message":"Executing (default): 
     SELECT \"id\", ROUND(AVG(\"rating\"), 2) AS \"ratingAvg\", \"serviceProviderId\" 
     FROM \"rating\" AS \"Rating\" 
     WHERE (\"Rating\".\"serviceProviderId\" IN (2, 1) AND \"Rating\".\"isFromUser\" = true) 
     GROUP BY \"serviceProviderId\";","level":"info"}

error: column "Rating.id" must appear in the GROUP BY clause or be used in an aggregate function
```


Since I only did an aggregation, I thought I wouldn't need to group by the "id" (Rating.id) column neither put any columns in the exclude attribute.
So I put the "id" in the group by clause to test.

The code is now this:

```javascript
// other includes... 
{
  model: Rating,
  as: `rating`,
  separate: true,
  attributes: [
	[
	  sequelize.fn(`ROUND`, sequelize.fn(`AVG`, sequelize.col(`rating`)), 2), `ratingAvg`,
	],
  ],
  where: { isFromUser: true },
  group: [`id`, `serviceProviderId`], // changed line
}
```

Here's the SQL output:

```SQL
[winston] Attempt to write logs with no transports {"message":"Executing (default): 
     SELECT \"id\", ROUND(AVG(\"rating\"), 2) AS \"ratingAvg\", \"serviceProviderId\" 
     FROM \"rating\" AS \"Rating\" 
     WHERE (\"Rating\".\"serviceProviderId\" IN (2, 1) AND \"Rating\".\"isFromUser\" = true) 
     GROUP BY \"id\", \"serviceProviderId\";","level":"info"}
```

And here is the returned objects:

```javascript
rating:  [
  Rating {
    dataValues: { ratingAvg: '5.00' },
    _previousDataValues: { ratingAvg: '5.00' },
    _changed: {},
    // other stuffs...
  },
  Rating {
    dataValues: { ratingAvg: '2.00' },
    _previousDataValues: { ratingAvg: '2.00' },
    _changed: {},
	// other stuffs...
  }
]
```

An array of the ratings (an object to each serviceProviderId record). Those values are right indeed, but it should have **calculated the average and returned me an object.**


So I tried the most "hard-worked way": I put it all the columns I don't want to be in the select script in the exclude attribute.


The new code:

```javascript
{
  model: Rating,
  as: `rating`,
  separate: true,
  attributes: {
	include: [
	  [
		sequelize.fn(`ROUND`, sequelize.fn(`AVG`, sequelize.col(`rating`)), 2), `ratingAvg`,
	  ],
	],
	exclude: [
	  `id`,
	  `userId`,
	  `serviceProviderId`,
	  `rating`,
	  `isFromUser`,
	  `createdAt`,
	  `updatedAt`,
	  `deletedAt`,
	],
  },
  where: { isFromUser: true },
  group: [`serviceProviderId`],
}
```


**Note**: see that now the "attribute" attribute is no longer an array but an object as the Sequelize docs suggest when we want to work with both include and exclude fields (https://sequelize.org/master/manual/model-querying-basics.html#specifying-attributes-for-select-queries)


And here is the new SQL output:

```SQL
[winston] Attempt to write logs with no transports {"message":"Executing (default): 
     SELECT \"id\", ROUND(AVG(\"rating\"), 2) AS \"ratingAvg\", \"serviceProviderId\" 
     FROM \"rating\" AS \"Rating\" 
     WHERE (\"Rating\".\"serviceProviderId\" IN (2, 1) AND \"Rating\".\"isFromUser\" = true) 
     GROUP BY \"serviceProviderId\";","level":"info"}

error: column "Rating.id" must appear in the GROUP BY clause or be used in an aggregate function
```

And even If I put the "id" back to the group by clause, it continues to output the same array of values.



Here is the SQL script I'm running on my own:

```SQL
     SELECT ROUND(AVG("rating"), 2) AS "ratingAvg"
     FROM "rating"
     WHERE "serviceProviderId" IN (2, 1) AND "isFromUser" = true
     GROUP BY "serviceProviderId"
```

It returns this:

![image](https://user-images.githubusercontent.com/17495320/100660445-cb77b980-3330-11eb-88ea-3ce4a60ef84a.png)

3.5 is the correct value.

To sum up, this is what I'm trying to achieve via my code. I'm fraking out :sweat_smile:

**Comment by Schukuratsu** [comment]
You can solve this easily using a subquery, check it here: https://sequelize.org/master/manual/sub-queries.html

**Comment by lnickersl** [comment]
Had same problem with this:

`const product = await Product.findOne({
  where: {id},
  include: [
    {
      model: Rating,
      as: 'ratings',
      attributes: [],
    },
  ],
  attributes: {
    include: [
      [sequelize.fn('AVG', sequelize.col('ratings.rating')), 'avgRating'],
    ],
  },
  group: ['Product.id', 'ratings.id'],
})`
    
**You have 2 options:**
1. Remove "attributes: []" and have Ratings attached to Product body
2. Remove "ratings.id" from "group" and have AVG work properly

I don't know how to get both, but luckely I don't need Ratings fields
